Docs M2

Cam0Cow opened the pull request:
# Quality Assurance Checklist 
To make reviews more efficient, please make sure the software feature meets the following standards and check everything off that meets the quality check. Once everything has been checked, the assigned reviewers will begin the review process. _Edit this description to check off the list._

There are exceptions with all guidelines. As long as your decisions are justified, then you are good! Contact the reviewers or the leads about any exceptions. 

## Requirements 
- [x] Followed Coding Style Guide 
- [ ] Presented/discussed in some capacity with others on the Controls team
- [x] Code Build checks pass 
- [x] No merge conflicts 
- [x] Software feature has documentation for it updated in both ReadTheDocs and the comments
- [x] Software feature has documentation for it updated
- Testing
    - [ ] Software feature has test associated with it
    - [ ] Test provides useful information and uses relevant data to accurately represent Controls 
    - [ ] Tested on hardware
    - NOTE: If test file already exists, use that one
- [x] If applicable, have you added the new feature to main.c?
- [x] Tagged appropriate issue ticket on this PR
- [ ] Did you have fun?

## Things to Consider 
- [ ] Even if the above are checked, is this the best way of writing my code? 
    - It's possible to write code that works, but are there ways to make code more efficient?

Cam0Cow committed:
Update requirements.txt and overview

Cam0Cow committed:
Fix up BSP layer docs

Cam0Cow committed:
Update conf and landing page

Cam0Cow committed:
Modify drivers docs

Cam0Cow committed:
Actually modified drivers docs

IshDeshpa committed:
Merge branch 'master' into docs-m2

IshDeshpa committed:
fixed up docs for Main and ReadCarCAN

IshDeshpa committed:
moved macros from ReadCarCAN.c to ReadCarCAN.h so they are visible in docs

IshDeshpa committed:
fixed header comment so doxygen recognizes it

IshDeshpa committed:
created helper script for printing all sphinx references

IshDeshpa committed:
updated readtritium docs

IshDeshpa committed:
updated Minions.rst

IshDeshpa commented:
Task linked: [CU-868680632 Update Documentation for M2.0](https://app.clickup.com/t/868680632)

IshDeshpa committed:
stash pop

IshDeshpa committed:
small changes fixing sphinx warnings and other clerical issues

IshDeshpa committed:
changed all doxygengroups to doxygenfiles and removed all doxygengroups entirely

IshDeshpa committed:
ReadCarCAN complete

IshDeshpa committed:
documented SendTritium.h

IshDeshpa committed:
added brief to SendTritium, fixed small build error in ReadCarCAN

IshDeshpa committed:
updated UpdateDisplay docs

IshDeshpa committed:
SendCarCAN updates

IshDeshpa committed:
updated ReadTritium docs

IshDeshpa committed:
almost finished with updating Apps layer docs

IshDeshpa committed:
more updates to SendTritium docs

IshDeshpa committed:
build fix

IshDeshpa committed:
changed ordering

IshDeshpa committed:
More workflow updates lol. Now, all documentation will be included within the
source code, and will be generated using Doxygen. This will allow us to
keep the documentation contained within the source code, and will make it
easier to keep the documentation up to date. Transition process has already started.

IshDeshpa committed:
more updates (trying to figure out images)

IshDeshpa committed:
configured more breathe stuff, finally figured out how to do images from markdown, updated FaultState with new ErrorHandling information

IshDeshpa committed:
more docs updates

IshDeshpa committed:
Contactors documentation updates

IshDeshpa committed:
finished last bits of driver documentation

IshDeshpa committed:
Merge branch 'master' into docs-m2

IshDeshpa committed:
added BSP and StdPeriph docs

IshDeshpa committed:
docs build with a lot of warnings, but they're all from StdPeriph so I assume it's fine

IshDeshpa committed:
Merge branch 'master' into docs-m2

IshDeshpa commented:
Run `make docs` in the Controls top-level folder. Then view the output using this [extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode.live-server) by navigating to the Controls/Docs/build/html directory.

NathanielDelgado commented on Apps/Inc/SendTritium.h:
```diff
@@ -1,52 +1,102 @@
-    BRAKE_STATE,
-    ONEPEDAL,
-    ACCELERATE_CRUISE
+    /** Forward drive state */
+    FORWARD_DRIVE,   
+    /** Neutral drive state */
+    NEUTRAL_DRIVE,   
+    /** Reverse drive state */
+    REVERSE_DRIVE,   
+    /** Record velocity state (cruise) */
+    RECORD_VELOCITY, 
+    /** Powered cruise state (cruise) */
+    POWERED_CRUISE,  
+    /** Coasting cruise state (cruise) */
+    COASTING_CRUISE, 
+    /** Brake state */
+    BRAKE_STATE,     
+    /** One pedal state (regen) */
+    ONEPEDAL,        
+    /** Accelerate cruise state (cruise) */
+    ACCELERATE_CRUISE   
```
Not sure if this type of commenting is neccessary just because it can feel a bit redundant but doesn't matter too much. Might be better to over document then under.

IshDeshpa replied:
Good stuff man! Keep it up!

NathanielDelgado reviewed (changes requested):
![image](https://github.com/lhr-solar/Controls/assets/89701060/a591e6e6-bfe8-452a-a377-c094af76aad9)
![image](https://github.com/lhr-solar/Controls/assets/89701060/e0e8f7ed-de3e-4630-94dd-e2b58d82c242)
Delete the brief and def flags in the docs because they are unnecessary. Sometimes you might want to use brief but not always if there is not much detail to begin with

![image](https://github.com/lhr-solar/Controls/assets/89701060/44c8826c-81cd-4e5e-814a-7ae4f922f510)
Reorganize header and source file doxygen comments to have this placement and level of detail

Add the copywrite back

aidenmadaffri commented:
Getting an error on the build of the docs:
```
no theme named 'sphinx_book_theme' found (missing theme.conf?)
make[1]: *** [Makefile:19: html] Error 2
make[1]: Leaving directory '/home/aiden/LHR/Controls/Docs'
make: *** [Makefile:45: docs] Error 2
```

This was tested on a fresh Ubuntu install and pull of the repo.

NathanielDelgado commented:
> Getting an error on the build of the docs:
> 
> ```
> no theme named 'sphinx_book_theme' found (missing theme.conf?)
> make[1]: *** [Makefile:19: html] Error 2
> make[1]: Leaving directory '/home/aiden/LHR/Controls/Docs'
> make: *** [Makefile:45: docs] Error 2
> ```
> 
> This was tested on a fresh Ubuntu install and pull of the repo. Install script from Embedded Sharepoint was ran without any apparent errors.

@aidenmadaffri it is telling you you need to install a theme for sphinx to display. This is the theme that is used to generate the HTML documents. You can resolve this using `pip install sphinx_book_theme`. However, this should be fixed in the install script so maybe make an issue ticket for that? @IshDeshpa

aidenmadaffri commented:
> > Getting an error on the build of the docs:
> > ```
> > no theme named 'sphinx_book_theme' found (missing theme.conf?)
> > make[1]: *** [Makefile:19: html] Error 2
> > make[1]: Leaving directory '/home/aiden/LHR/Controls/Docs'
> > make: *** [Makefile:45: docs] Error 2
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This was tested on a fresh Ubuntu install and pull of the repo. Install script from Embedded Sharepoint was ran without any apparent errors.
> 
> @aidenmadaffri it is telling you you need to install a theme for sphinx to display. This is the theme that is used to generate the HTML documents. You can resolve this using `pip install sphinx_book_theme`. However, this should be fixed in the install script so maybe make an issue ticket for that? @IshDeshpa

This resolves the issue, but yes it should be fixed in the install script.

aidenmadaffri commented on Apps/Inc/ReadTritium.h:
```diff
@@ -15,29 +14,48 @@
     T_HARDWARE_OVER_CURRENT_ERR = (1<<0), 
-    T_SOFTWARE_OVER_CURRENT_ERR = (1<<1), 
-    T_DC_BUS_OVERVOLT_ERR = (1<<2), 
+    /** software over current error */
```
This is minor, but the rest of the files have comments starting with a capital letter. These comments should probably be updated to reflect that.

aidenmadaffri reviewed (dismissed):
Very clear explanation of all the files and their functions. This will be an excellent resource to reference.

KnockbackNemo commented on Apps/Inc/ReadTritium.h:
```diff
@@ -1,10 +1,9 @@
- * @brief 
+ * @brief In its current iteration, the Read Tritium task facilitates all received
+ * communication from MotorCAN. All messages are forwarded to CarCAN, and velocity and error
+ * information are recorded.
```
Might be nice to note where the info is recorded and how it gets accessed

KnockbackNemo commented on Apps/Inc/SendCarCAN.h:
```diff
@@ -1,3 +1,11 @@
+/**
+ * @file SendCarCAN.h
+ * @brief The Send Car CAN task is a simple queue consumer task. Multiple 
+ * tasks that need to write the the car CAN bus; in order to do this safely, 
```
```suggestion
 * tasks that need to write to the car CAN bus; in order to do this safely, 
```

KnockbackNemo commented on Apps/Src/ReadCarCAN.c:
```diff
@@ -1,7 +1,17 @@
+ * We also expect to receive supplemental voltage and state of charge messages from BPS. Both of these are simply stored to show on the display.
+ * 
+ * # Implementation Details
+ * The Read Car CAN task uses RTOS timers to make sure message timings remain appropriate. A watchdog is pet on BPS contactor state message receive to ensure that if communication is lost with BPS, the system disables the contactors.
```
```suggestion
 * The Read Car CAN task uses RTOS timers to make sure message timings remain appropriate. A watchdog is pet on receiving the BPS contactor state message to ensure that if communication is lost with BPS, the system disables the contactors.
```

KnockbackNemo commented on Apps/Src/ReadTritium.c:
```diff
@@ -1,10 +1,16 @@
+ * read out by the SendCarCAN task.
+ * 
+ * ReadTritium also facilitates reading velocity and error information in from the motor controller. 
+ * Velocity is put on the display, and error information is used to determine if the motor controller 
+ * is in a fault state. If the motor controller is in a fault state, the task will attempt to reset the 
+ * motor controller. If the motor controller cannot be reset, the task will set the car to a fault state.
```
Currently, most errors (except hall sensor and motor watchdog) will result in a nonrecoverable fault without any attempts to reset the motor controller, so perhaps we could write something like this
```suggestion
 * Velocity is put on the display, and error information is used to determine if the motor controller 
 * has detected an error. Depending on the frequency and severity of the error, the task will either 
 * attempt to reset the motor controller or set the car to a fault state.
```

KnockbackNemo commented on Apps/Src/UpdateDisplay.c:
```diff
@@ -1,11 +1,14 @@
- * relevant information about system status to the driver.
+ * @details
+ * The task implements a queue of command structures. Tasks that wish to 
+ * write the display call one of the wrapper functions exposed in the public interface. 
```
```suggestion
 * write to the display call one of the wrapper functions exposed in the public interface. 
```

KnockbackNemo commented on BSP/Inc/BSP_CAN.h:
```diff
@@ -30,27 +31,25 @@ typedef enum {CAN_1=0, CAN_3, NUM_CAN} CAN_t;
+ * @note    Non-blocking statement
+ * @pre     The data parameter must be at least 8 bytes or hardfault may occur.
+ * @param   id : pointer to store id of the message that was received.
+ * @param   data : pointer to store data that was received. Must be 8bytes or bigger.
```
```suggestion
 * @param   data : pointer to store data that was received. Must be 8 bytes or bigger.
```

KnockbackNemo commented on BSP/STM32F413/Src/BSP_GPIO.c:
```diff
@@ -1,4 +1,7 @@
-/* Copyright (c) 2021 UT Longhorn Racing Solar */
+/**
+ * @file    BSP_GPIO.c
+ * @details
```
Just wanted to check if we plan to leave this empty

KnockbackNemo commented on BSP/STM32F413/Src/BSP_GPIO.c:
```diff
@@ -66,12 +69,11 @@ void BSP_GPIO_Write(port_t port, uint16_t data){
 
 /**
- * @brief   Reads data from a specified input pin (not applicable to output pins)
+ * @brief   Reads data from a specified pin (not applicalbe to output pins)
```
```suggestion
 * @brief   Reads data from a specified pin (not applicable to output pins)

```

KnockbackNemo commented on BSP/STM32F413/Src/BSP_GPIO.c:
```diff
@@ -83,11 +85,10 @@ uint8_t BSP_GPIO_Read_Pin(port_t port, uint16_t pinmask){
  * @brief   Writes data to a specified pin
  * @param   port The port to write to
- * @param   pin The pin to write to 
+ * @param   pinmask Mask from stm header file that says which pin to write too
```
```suggestion
 * @param   pinmask Mask from stm header file that says which pin to write to

```

KnockbackNemo commented on BSP/STM32F413/Src/BSP_UART.c:
```diff
@@ -192,10 +207,12 @@ uint32_t BSP_UART_Read(UART_t usart, char *str) {
- * @param   str : pointer to buffer with data to send.
- * @param   len : size of buffer
- * @param   usart : which usart to read from (2 or 3)
+ * @brief   Transmits data to through a specific 
```
```suggestion
 * @brief   Transmits data through a specific 
```

KnockbackNemo commented on Docs/source/index.rst:
```diff
@@ -3,65 +3,63 @@
-
-   Overview/Hardware
-   Overview/Software
+To actually build the documentation, run ``make docs``. The resulting ``build/html/index.html`` can then be viewed in any browser. When your pull request is merged into the master branch, the documentation changes should be automatically visible on this website.
```
If we're recommending the extension, then we could consider mentioning it here too
```suggestion
To actually build the documentation, run ``make docs``. The resulting ``build/html/index.html`` can then be viewed in any browser or in the editor using this [VSCode extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode.live-server) by navigating to the Controls/Docs/build/html directory. When your pull request is merged into the master branch, the documentation changes should be automatically visible on this website.
```

KnockbackNemo commented on Drivers/Inc/CANbus.h:
```diff
@@ -64,55 +64,57 @@ typedef struct {
+ * @brief   Reads a CAN message from the CAN hardware and returns it to the provided pointers
+ * @param   data 		where to store the received message
+ * @param   blocking 	Whether or not this read should be blocking
+ * @param   bus 		The bus to use, either be CARCAN or MOTORCAN
```
```suggestion
 * @param   bus 		The bus to use, can either be CARCAN or MOTORCAN
```

KnockbackNemo commented on Drivers/Inc/Minions.h:
```diff
@@ -49,22 +57,20 @@ void Minions_Init(void);
- * @param status 
- * @return true is fail (wrote to an input)
- * @return false is success (wrote to an output)
+ * @param pin the pint to read
```
```suggestion
 * @param pin the pin to read
```

KnockbackNemo reviewed (changes requested):
Just a few small proofreading suggestions to support all the great work already done.

IshDeshpa committed:
fixed ReadTritium

diyarajon commented on Docs/source/Apps/ErrorHandling.rst:
```diff
@@ -0,0 +1,50 @@
+===========
+All OS functions return an error code whose value should immediately be checked by a call to ``assertOSError``. The Controls system cannot run without a properly functioning OS, so any OS error is critical and nonrecoverable. Thus, when an OS function returns an error code that is not ``OS_ERR_NONE``,  the assertion will fail, and the following steps will occur:
+    #. If in DEBUG mode, the file and line of the failed assertion will be printed
+    #. The Array Bypass-Precharge Contactor and Motor Bypass-Precharge Contactor will open
```
```suggestion
    #. The Array Precharge-Bypass Contactor and Motor Precharge-Bypass Contactor will open
```

diyarajon commented on Docs/source/Apps/ErrorHandling.rst:
```diff
@@ -0,0 +1,50 @@
+The errors asserted by applications include
+
+* :doc:`ReadTritium <ReadTritium>`: Motor controller errors
+    * Errors sent from the motor controller are asserted by ReadTritium. These are typically nonrecoverable, with the exception of the motor watchdog and hall sensor errors.
```
```suggestion
    * Errors sent from the motor controller are asserted by ReadTritium. These are typically nonrecoverable, with the exception of the motor watchdog and hall sensor errors. Motor watchdog and hall sensor errors attempt to recover a set number of times before entering a nonrecoverable state. 
```

diyarajon commented on Docs/source/Apps/ErrorHandling.rst:
```diff
@@ -0,0 +1,50 @@
+* :doc:`ReadTritium <ReadTritium>`: Motor controller errors
+    * Errors sent from the motor controller are asserted by ReadTritium. These are typically nonrecoverable, with the exception of the motor watchdog and hall sensor errors.
+* :doc:`ReadCarCAN <ReadCarCAN>`: Contactor disable conditions
+    * These errors depend on BPS and include charge disable, contactor open, and BPS trip, as well as missed BPS messages.
```
```suggestion
    * These errors depend on BPS and include charge disable, contactor open, and BPS trip, as well as missed BPS messages. BPS trip is nonrecoverable, and all the other errors are recoverable. All the errors have callback functions. 
```

IshDeshpa committed:
more readtritium fixes

diyarajon reviewed (changes requested):
Looks great!

IshDeshpa committed:
Fixed ReadCarCAN

IshDeshpa committed:
updated SendTritium

IshDeshpa committed:
updated UpdateDisplay docs

IshDeshpa committed:
updates to naming

NathanielDelgado committed:
Dev/code cleanup again (#404)

* Added clang format/tidy config files, changes to the makefiles to apply them and edited the github workflow to make sure the tools are enforced

* Applied clang format and tidy

* Updated embedded shairpoint which has correct install script

IshDeshpa committed:
update sendcarcan docs

IshDeshpa committed:
misc changes

NathanielDelgado committed:
Modifed workflow to enforce clang and format (#408)

IshDeshpa committed:
Merge remote-tracking branch 'origin/master' into docs-m2 (This took me like three hours)

IshDeshpa committed:
Merge branch 'master' into docs-m2

IshDeshpa committed:
make format fixes

IshDeshpa committed:
applied clang tidy/format fixes and fixed build after merge

IshDeshpa committed:
renames

IshDeshpa committed:
updated more naming, updated Tasks.c and added diagram for the TaskSwHook

IshDeshpa committed:
got error handling to show up as a sub item of tasks, renamed Main to Initialization (only in docs)

diyarajon commented on Apps/Inc/SendCarCan.h:
```diff
@@ -1,3 +1,14 @@
+/**
+ * @file SendCarCAN.h
```
```suggestion
 * @file SendCarCan.h
```

diyarajon commented on Apps/Inc/SendCarCan.h:
```diff
@@ -1,3 +1,14 @@
+/**
+ * @file SendCarCAN.h
+ * @brief Sends any CarCAN messages put in the queue via SendCarCAN_Put, and handles
```
```suggestion
 * @brief Sends any CarCAN messages put in the queue via SendCarCanPut, and handles
```

diyarajon commented on Apps/Inc/SendCarCan.h:
```diff
@@ -1,3 +1,14 @@
+ * @brief Sends any CarCAN messages put in the queue via SendCarCAN_Put, and handles
+ * transmitting the IO State of the car for telemetry & ignition sequence.
+ * 
+ * Call SendCarCAN_Init() to initialize the queue and semaphore. Call SendCarCAN_Put() to 
```
```suggestion
 * Call SendCarCanInit() to initialize the queue and semaphore. Call SendCarCanPut() to 
```

diyarajon commented on Apps/Inc/SendCarCan.h:
```diff
@@ -1,3 +1,14 @@
+/**
+ * @file SendCarCAN.h
+ * @brief Sends any CarCAN messages put in the queue via SendCarCAN_Put, and handles
+ * transmitting the IO State of the car for telemetry & ignition sequence.
```
```suggestion
 * transmitting the IO State of the car for telemetry & ignition state.
```

diyarajon commented on Apps/Inc/SendCarCan.h:
```diff
@@ -1,3 +1,14 @@
+ * transmitting the IO State of the car for telemetry & ignition sequence.
+ * 
+ * Call SendCarCAN_Init() to initialize the queue and semaphore. Call SendCarCAN_Put() to 
+ * put a message in the queue. Starting the SendCarCAN task will send any messages
```
```suggestion
 * put a message in the queue. Starting the SendCarCan task will send any messages
```

diyarajon commented on Apps/Inc/SendCarCan.h:
```diff
@@ -10,13 +21,15 @@ void SendCarCanInit();
- */
 #ifdef DEBUG
+/**
+ * @brief Return the space left in SendCarCAN_Q for debug purposes
+ * @return the space left in SendCarCAN_Q
```
Do we use SendCarCAN_Q naming?

diyarajon reviewed (changes requested):
Looks good

IshDeshpa committed:
minor edits

IshDeshpa committed:
sendcarcan
